denonavr/api.py

```diff
@@ -0,0 +1,286 @@
+        default=httpx.Timeout(2.0, read=15.0))
+    _appcommand_update_tags: Tuple[AppCommandCmd] = attr.ib(
+        validator=attr.validators.deep_iterable(
+            attr.validators.instance_of(AppCommandCmd),
+            attr.validators.instance_of(tuple)),
+        default=attr.Factory(tuple))
+    _appcommand0300_update_tags: Tuple[AppCommandCmd] = attr.ib(
+        validator=attr.validators.deep_iterable(
+            attr.validators.instance_of(AppCommandCmd),
+            attr.validators.instance_of(tuple)),
+        default=attr.Factory(tuple))
+
+    def __hash__(self) -> int:
+        """
+        Hash the class in a custom way that caching works.
+
+        It should react on changes of host and port.
+        """
+        return hash((id(self), self.host, self.port))
+
+    @async_handle_receiver_exceptions
+    async def async_get(
+            self,
+            request: str,
+            port: Optional[int] = None) -> httpx.Response:
+        """Call GET endpoint of Denon AVR receiver asynchronously."""
+        # Use default port of the receiver if no different port is specified
+        port = port if port is not None else self.port
+
+        endpoint = "http://{host}:{port}{request}".format(
+            host=self.host, port=port, request=request)
+
+        async with httpx.AsyncClient() as client:
+            res = await client.get(endpoint, timeout=self.timeout)
+            res.raise_for_status()
+
+        return res
+
+    @async_handle_receiver_exceptions
+    async def async_post(
+            self,
+            request: str,
+            content: Optional[bytes] = None,
+            data: Optional[Dict] = None,
+            port: Optional[int] = None) -> httpx.Response:
+        """Call POST endpoint of Denon AVR receiver asynchronously."""
+        # Use default port of the receiver if no different port is specified
+        port = port if port is not None else self.port
+
+        endpoint = "http://{host}:{port}{request}".format(
+            host=self.host, port=port, request=request)
+
+        async with httpx.AsyncClient() as client:
+            res = await client.post(
+                endpoint, content=content, data=data, timeout=self.timeout)
+            res.raise_for_status()
+
+        return res
+
+    @async_handle_receiver_exceptions
+    async def async_get_command(self, request: str) -> str:
+        """Send HTTP GET command to Denon AVR receiver asynchronously."""
+        # HTTP GET to endpoint
+        res = await self.async_get(request)
+        # Return text
+        return res.text
+
+    @set_cache_id
+    @alru_cache(maxsize=32, cache_exceptions=False)
+    @async_handle_receiver_exceptions
+    async def async_get_xml(
+            self,
+            request: str,
+            cache_id: Hashable = None
+            ) -> ET.Element:
+        """Return XML data from HTTP GET endpoint asynchronously."""
+        # HTTP GET to endpoint
+        res = await self.async_get(request)
```

Comment:
Why not use `async_get_command`?

Reply:
One way or the other, this make not too much difference
I built `async_get_command` especially to send commands like in the previous version. I just return the output, because I've seen a pull request in home-assistant to handle the returned data.